### PR TITLE
Collapse data convert to multiple 1-line conversions

### LINKED ISSUE #302: Consider breaking up the data converter
Look into the feasibility of breaking up the `ingest_info proto` -> `persistence layer entities` converter which will help isolate code (prevent lots of logic in one file) & make it easier to write more in depth tests (eg. comprehensive tests for parsing a booking). We should also write those more detailed tests.

### LINKED ISSUE #316: Make the converter pre-process the proto
We want the converter to pre-process the proto into a mapping from entity id to the object.  This will make the converter code much easier to read.  We can also consider just using maps in the proto which would solve this problem.

### PR DESCRIPTION
**Pull request by zacharylawrence**
Collapse data conversion by wrapping checking if the proto has a field in the `fn` method.

Fixes #316
Fixes #302

### CODE REVIEW

File: recidiviz/persistence/converter.py [hunk 1]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
-            arrest.date = self._convert(converter_utils.parse_date_or_error,
-                                        ingest_arrest.date)
+        return new
 
-        if ingest_arrest.HasField('location'):
-            arrest.location = self._convert(converter_utils.normalize,
-                                            ingest_arrest.location)
+    def _convert_booking(self, ingest_booking):
+        """Converts an ingest_info proto Booking to a persistence entity."""
+        p = ingest_booking
```
Comment (jessex) [inline recidiviz/persistence/converter.py]: Should this be `b` instead of `p`?

File: recidiviz/persistence/converter.py [hunk 2]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
-                                          ingest_charge.degree)
+        return new
 
-        if ingest_charge.HasField('charge_class'):
-            charge.charge_class = self._convert(
-                ChargeClass.from_str, ingest_charge.charge_class)
+    @staticmethod
+    def _convert_arrest(ingest_arrest):
+        """Converts an ingest_info proto Arrest to a persistence entity."""
+        p = ingest_arrest
```
Comment (jessex) [inline recidiviz/persistence/converter.py]: `a` instead of `p`?

File: recidiviz/persistence/converter.py [hunk 3]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
-                converter_utils.parse_dollar_amount,
-                ingest_charge.fee_dollars)
+        return arrest
 
-        if ingest_charge.HasField('charging_entity'):
-            charge.charging_entity = self._convert(
-                converter_utils.normalize, ingest_charge.charging_entity)
+    def _convert_charge(self, ingest_charge):
+        """Converts an ingest_info proto Charge to a persistence entity."""
+        p = ingest_charge
```
Comment (jessex) [inline recidiviz/persistence/converter.py]: `c` instead of `p`?

File: recidiviz/persistence/converter.py [hunk 4]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
-        if ingest_charge.HasField('sentence_id'):
-            charge.sentence = self._convert_sentence(
-                ingest_charge.sentence_id)
+        return new
 
-        return charge
+    @staticmethod
+    def _convert_bond(ingest_bond):
+        """Converts an ingest_info proto Bond to a persistence entity."""
+        p = ingest_bond
```
Comment (jessex) [inline recidiviz/persistence/converter.py]: `b` instead of `p`?

File: recidiviz/persistence/converter.py [hunk 5]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
-    return error_str
+        if new.status is None:
+            new.status = BondStatus.POSTED
+
+        return new
+
+    @staticmethod
+    def _convert_sentence(ingest_sentence):
+        """Converts an ingest_info proto Sentence to a persistence entity."""
+        p = ingest_sentence
```
Comment (jessex) [inline recidiviz/persistence/converter.py]: `s` instead of `p`?

File: recidiviz/persistence/converter.py [hunk 6]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
+        new.fine_dollars = fn(parse_dollars, 'fine_dollars', p)
+        new.parole_possible = fn(parse_bool, 'parole_possible', p)
+        new.post_release_supervision_length_days = \
+            fn(parse_days, 'post_release_supervision_length', p)
+
+        return new
+
+
+def fn(func, field_name, proto):
+    """Return the result of apply the given function to the field on the proto,
```
Comment (jessex) [inline recidiviz/persistence/converter.py]: apply --> applying?

**Comment by jessex** [comment]
Only nits from me. Great work, @zacharylawrence! Such a significant readability improvement.

File: recidiviz/persistence/converter.py [hunk 7]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
+                # TODO(338): Clean up the default logic here.
+                inferred_charge.status = ChargeStatus.PENDING
 
-        if ingest_charge.HasField('name'):
-            charge.name = self._convert(converter_utils.normalize,
-                                        ingest_charge.name)
+                new.charges = [inferred_charge]
+            else:
+                if any(charge.bond is not None for charge in new.charges):
+                    raise ValueError(
```
Comment (jamwalla) [inline recidiviz/persistence/converter.py]: 🙏🏽

File: recidiviz/persistence/converter.py [hunk 8]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
-                converter_utils.normalize, ingest_charge.charging_entity)
+    def _convert_charge(self, ingest_charge):
+        """Converts an ingest_info proto Charge to a persistence entity."""
+        p = ingest_charge
+        new = entities.Charge()
+
+        new.external_id = fn(normalize, 'charge_id', p)
+        new.offense_date = fn(parse_date, 'offense_date', p)
+        new.statute = fn(normalize, 'statute', p)
+        # TODO(215): charge.offense_code = code_to_BJS(ingest_charge.statute)
```
Comment (jamwalla) [inline recidiviz/persistence/converter.py]: can you remove this line? (not related to your change)

File: recidiviz/persistence/converter.py [hunk 9]
```diff
@@ -46,399 +44,208 @@ class _Converter(object):
...
-        return arrest
+        new.facility = fn(normalize, 'facility', p)
+        new.classification = fn(Classification.from_str, 'classification', p)
 
-    def _convert_charge(self, charge_id):
-        """Converts an ingest_info proto Charge to a persistence object."""
-        ingest_charge = more_itertools.one(
-            charge for charge in self.ingest_info.charges if
-            charge.charge_id == charge_id)
+        new.arrest = fn(lambda i: _Converter._convert_arrest(self.arrests[i]),
```
Comment (jamwalla) [inline recidiviz/persistence/converter.py]: If we make `fn` a method on `Converter`, does it work to change this to `self._convert_arrest` (and same for bond)? Feels weird to have some of these be static and others not. But if you think that's worse I see that too!

File: recidiviz/persistence/converter.py (hunk 1 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: I had intended this to mean `proto` but that's clear now after breaking things out into new files.

File: recidiviz/persistence/converter.py (hunk 9 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: After breaking this up into multiple files, I think this is resolved. Feel free to open it back up if you disagree!

File: recidiviz/persistence/converter.py (hunk 2 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: I had intended this to mean `proto` but that's clear now after breaking things out into new files.

File: recidiviz/persistence/converter.py (hunk 3 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: I had intended this to mean `proto` but that's clear now after breaking things out into new files.

File: recidiviz/persistence/converter.py (hunk 8 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: Done.

File: recidiviz/persistence/converter.py (hunk 4 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: I had intended this to mean `proto` but that's clear now after breaking things out into new files.

File: recidiviz/persistence/converter.py (hunk 5 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: I had intended this to mean `proto` but that's clear now after breaking things out into new files.

File: recidiviz/persistence/converter.py (hunk 6 above)
Comment (zacharylawrence) [reply recidiviz/persistence/converter.py]: Nice catch, done.

File: recidiviz/persistence/converter/converter_utils.py [hunk 10]
```diff
@@ -15,14 +15,22 @@
...
 import datetime
 from distutils.util import strtobool  # pylint: disable=no-name-in-module
 
 from recidiviz.common import common_utils
 from recidiviz.common.constants.person import Ethnicity, Race
 
 
+def fn(func, field_name, proto, default=None):
+    """Return the result of applying the given function to the field on the
+    proto, returning None if the proto field is unset.
```
Comment (jessex) [inline recidiviz/persistence/converter/converter_utils.py]: "... returning |default| if the proto field is unset."

Comment (zacharylawrence) [reply recidiviz/persistence/converter/converter_utils.py]: Nice catch! Done.

**Comment by zacharylawrence** [comment]
Thanks for both of you taking a 2nd pass on this one!

**Pull request merged**